src/worker/activities/activity_heartbeat_manager.rs

```diff
@@ -105,143 +111,221 @@ impl ActivityHeartbeatManager {
+            self.throttle_interval - time_since_last_sent
+        } else {
+            Duration::ZERO
+        };
     }
 }
 
 #[derive(Debug)]
-struct ActivityHbState {
-    delay: Duration,
-    last_sent: Instant,
+struct HeartbeatStreamState {
+    tt_to_state: HashMap<TaskToken, ActivityHeartbeatState>,
+    incoming_hbs: UnboundedReceiver<HeartbeatAction>,
+    /// Token that can be used to cancel the entire stream.
+    /// Requests to the server are not cancelled with this token.
+    pub cancellation_token: CancellationToken,
+}
+
+impl HeartbeatStreamState {
+    fn new() -> (Self, UnboundedSender<HeartbeatAction>) {
+        let (heartbeat_tx, incoming_hbs) = unbounded_channel();
+        (
+            Self {
+                cancellation_token: Default::default(),
+                tt_to_state: Default::default(),
+                incoming_hbs,
+            },
+            heartbeat_tx,
+        )
+    }
+
+    /// Record a heartbeat received from lang
+    fn record(&mut self, hb: ValidActivityHeartbeat) -> Option<HeartbeatExecutorAction> {
+        match self.tt_to_state.entry(hb.task_token.clone()) {
+            Entry::Vacant(e) => {
+                let state = ActivityHeartbeatState {
+                    task_token: hb.task_token.clone(),
+                    throttle_interval: hb.throttle_interval,
+                    last_send_requested: Instant::now(),
+                    last_recorded_details: None,
+                    sleep_cancel_tx: None,
+                };
+                e.insert(state);
+                Some(HeartbeatExecutorAction::Report(hb.task_token, hb.details))
+            }
+            Entry::Occupied(mut o) => {
+                let state = o.get_mut();
+                state.last_recorded_details = Some(hb.details);
+                None
+            }
+        }
+    }
+
+    /// Heartbeat report to server completed
+    fn complete_report(&mut self, tt: TaskToken) -> Option<HeartbeatExecutorAction> {
+        if let Some(st) = self.tt_to_state.get_mut(&tt) {
+            let cancellation_token = self.cancellation_token.child_token();
+            st.sleep_cancel_tx = Some(cancellation_token.clone());
+            // Always sleep for simplicity even if the duration is 0
+            Some(HeartbeatExecutorAction::Sleep(
+                tt.clone(),
+                st.get_throttle_sleep_duration(),
+                cancellation_token,
+            ))
+        } else {
+            None
+        }
+    }
+
+    /// Throttling completed, report or stop tracking task token
+    fn complete_throttle(&mut self, tt: TaskToken) -> Option<HeartbeatExecutorAction> {
+        match self.tt_to_state.entry(tt.clone()) {
+            Entry::Occupied(mut e) => {
+                let state = e.get_mut();
+                if state.last_recorded_details.is_none() {
+                    // Nothing to report, forget this task token
+                    e.remove();
+                    None
+                } else {
```

Comment:
These branches could be simplified by having the first one be

`if let Some(details) = state.last_recorded_details.take() {`

Reply:
You're right.